### DISCUSSION TITLE
Randomness of MIlvus Collection

### DISCUSSION THREAD

**Discussion by Chahnwoo** [milvus-io/milvus#32733] [Q&A and General discussion] [answered]
I am working with the LangChain integration of Milvus and wanted to ask about how the Milvus HNSW algorithm works. Is there any degree of randomness introduced at any point of graph construction or search?

Here is a general overview of the collection generation process that I am using:

1. Retrieve a list of texts
2. Iteratively upload those texts into the MILVUS collection with a pre-defined embedding model

I am not using any randomization in the insert process and have been using the same embedding model and question set for all tests, yet seem to be getting different search results each time I create a new collection. Does anyone know what the problem might be?

**Comment by yhmo** [comment]
Some questions:
1. How many texts are uploaded to the Milvus collection?
2. Did you specify the "index_params" when you initialize the Milvus VectorStore?

**Reply by Chahnwoo** [reply]
1. I've uploaded approximately 80k LangChain Documents to the Milvus collection.
2. Yes, I did specify the index_params when I initialized the Milvus VectorStore. 

I've attached an image of part of the source code below.

![image](https://github.com/milvus-io/milvus/assets/91393402/2cc2937a-87e8-477f-ac73-05b2be064eb9)

**Reply by yhmo** [reply]
“different search results” --------------- The result is only a bit different, right?

HNSW is a multi-layer graph index, it randomly picks some vectors to the next layer when building the index. So, the search result might be a bit different for two HNSW indexes.

**Reply by Chahnwoo** [reply]
Ah, I see, thank you for that insight! Is there any way to control that randomness (with a seed or such), or will I have to use it as is?
Also, about the index that the HNSW algorithm uses: is it possible to construct an HNSW index with one metric_type (such as cosine) and then perform search with another (such as l2)?

**Reply by Chahnwoo** [reply]
@yhmo I have also been getting different search results for the same query, with the same Milvus collection. 
Here are some details about the process I have been using:

1. Initialize Milvus client
2. Choose a question
3. Repeatedly query the client with that question

While the retrieved documents are usually consistent, there are occasional cases in which the results differ.
Does this also have to do with the random selection of vectors to the next layer?

If this is the case, would increasing the 'ef_search' parameter of the index_params help?

**Reply by yhmo** [reply]
The description of HNSW: https://milvus.io/docs/index.md#HNSW
Index building parameters:
- **M**: defines tha maximum number of outgoing connections in the graph. Higher M leads to higher accuracy/run_time at fixed. Range: (2, 2048)
- **efConstruction**: controls index search speed/build speed tradeoff. Increasing the efConstruction parameter may enhance index quality, but it also tends to lengthen the indexing time.  Range: (1, int_max)

Search parameters:
**ef**: Parameter controlling query time/accuracy trade-off. Higher ef leads to more accurate but slower search. Range: [1, int_max]


To create a HNSW index:
```
 index_param = {"index_type": "HNSW", "metric_type": "L2", "params": {"M": 48, "efConstruction": 500}}
collection.create_index(vector_field, index_param)
```

To search:
```
search_param = {
        "data": search_vectors,
        "anns_field": VECTOR_FIELD_NAME,
        "param": {"metric_type": "L2", "params": {"ef": 64}},
        "limit": 10,
        "consistency_level": "Bounded",
    }
results = collection.search(**search_param)
```

The "ef" parameter should be input for search_param, not index_param.

You can increase the value of "ef" to observe the result.

**Reply by Chahnwoo** [reply]
Thank you for the answer! Can I take that to mean that there is no randomness introduced during the search process, the way it is during index building?

**Reply by yhmo** [reply]
Milvus splits data into segments, each segment has an independent index. Segment size could range from 100MB+ to 1GB+. Search engine searches topk results from each segment, and merges N topk to be a final topk result. So, different segment sizes slightly affect the search result. Segment size is affected by insert actions(batch by batch or row by row, insert interval, etc).
Index algorithm also affects search results.

**Reply by Chahnwoo** [reply]
But once the vector database has been fully created (no new inserts or deletions or updates take place), those same indices are used consistently for search, right? Assuming no modifications to a created vector database, there is no randomness introduced to search...?

**Reply by yhmo** [reply, marked as answer]
I mean two different full test processes might get different segment size.
Let's say we want to test 1M entities with different index types.
Process 1: create collection A, insert data batch by batch, 10000 entities for each batch. create index, test search
Process 2: create collection B, insert data batch by batch, 500 entities for each batch. create index, test search

Assume the total data size is 500MB.
The process 1 might get these segments: 130MB + 130MB + 130MB + 110MB.
The process 2 might get these segments: 120MB + 120MB + 120MB + 140MB

Data distribution is different. Search result might be a bit different.

If you use the same collection to test, no additional data inserted. Yes, segment size is not a problem.

**Reply by Chahnwoo** [reply]
Thank you so much! Your answers have been so helpful.
